agent/node.go

```diff
@@ -627,7 +627,19 @@ func (n *Node) runManager(ctx context.Context, securityConfig *ca.SecurityConfig
 			ElectionTick:   n.config.ElectionTick,
 		})
 		if err != nil {
-			return err
+			logrus.Errorf("error starting manager: %v", err)
+			// Something went wrong, but it might be fixed later on
+			// by user based on the error log. So rather than quit,
+			// we keep waiting for my role to be changed back if
+			// promotion failed. On the meantime, the agent part still
+			// work.
+			// TODO(jinuxstyle): change role back to agent if promotion
+			// failed in a self-healing way.
+			n.waitRole(ctx, ca.AgentRole)
+			if ctx.Err() != nil {
```

Comment:
I agree calling ctx.Err() after ctx.Done() is a good practice. But is it really not threadsafe if it's called before ctx.Done(). The context.Cancel() will set context.Err to non-nil and close context.Done immediately after that.